File: src/android/ScanditLabelCapture.kt
/*
 * This file is part of the Scandit Data Capture SDK
 *
 * Copyright (C) 2025- Scandit AG. All rights reserved.
 */

package com.scandit.datacapture.cordova.label

import com.scandit.datacapture.cordova.core.ScanditCaptureCore
import com.scandit.datacapture.cordova.core.utils.CordovaEventEmitter
import com.scandit.datacapture.cordova.core.utils.CordovaMethodCall
import com.scandit.datacapture.cordova.core.utils.CordovaResult
import com.scandit.datacapture.cordova.core.utils.PluginMethod
import com.scandit.datacapture.frameworks.core.CoreModule
import com.scandit.datacapture.frameworks.core.locator.DefaultServiceLocator
import com.scandit.datacapture.frameworks.label.LabelCaptureModule
import org.apache.cordova.CallbackContext
import org.apache.cordova.CordovaPlugin
import org.json.JSONArray
import org.json.JSONObject

class ScanditLabelCapture :
    CordovaPlugin() {

    private val emitter = CordovaEventEmitter()

    private val labelCaptureModule = LabelCaptureModule(emitter)

    private var lastLabelCaptureEnabledState: Boolean = false

    private val serviceLocator = DefaultServiceLocator.getInstance()

    override fun pluginInitialize() {
        super.pluginInitialize()
        ScanditCaptureCore.addPlugin(serviceName)
        labelCaptureModule.onCreate(cordova.context)
    }

    override fun onStop() {
        lastLabelCaptureEnabledState = labelCaptureModule.isTopmostModeEnabled()
        labelCaptureModule.setTopmostModeEnabled(false)
    }

    override fun onStart() {
        labelCaptureModule.setTopmostModeEnabled(lastLabelCaptureEnabledState)
    }

    override fun onReset() {
        labelCaptureModule.onDestroy()
        pluginInitialize()
    }

    override fun onDestroy() {
        labelCaptureModule.onDestroy()
        super.onDestroy()
    }

    override fun execute(
        action: String,
        args: JSONArray,
        callbackContext: CallbackContext
    ): Boolean {
        return when (action) {
            "executeLabel" -> executeLabel(args, callbackContext)
            "getDefaults" -> getDefaults(callbackContext)
            else -> false
        }.let { true }
    }

    fun executeLabel(args: JSONArray, callbackContext: CallbackContext) {
        val argsJson = args.getJSONObject(0)
        val coreModule = serviceLocator.resolve(
            CoreModule::class.java.simpleName
        ) as? CoreModule ?: return run {
            callbackContext.error("Unable to retrieve the CoreModule from the locator.")
        }

        val result = coreModule.execute(
            CordovaMethodCall(args),
            CordovaResult(callbackContext, emitter),
            labelCaptureModule
        )

        if (!result) {
            val methodName = argsJson.getString("methodName") ?: "unknown"
            callbackContext.error("Unknown method: $methodName")
        }
    }

    @PluginMethod
    fun getDefaults(callbackContext: CallbackContext) {
        callbackContext.success(
            JSONObject(
                mapOf(
                    "LabelCapture" to labelCaptureModule.getDefaults()
                )
            )
        )
    }
}

File: src/ios/ScanditLabelCapture.swift
import ScanditFrameworksCore
import ScanditFrameworksLabel

#if SWIFT_PACKAGE
import Cordova
import ScanditCordovaDatacaptureCore
#endif

@objc(ScanditLabelCapture)
public class ScanditLabelCapture: CDVPlugin {
    var labelModule: LabelCaptureModule!
    var emitter: CordovaEventEmitter!

    override public func pluginInitialize() {
        super.pluginInitialize()
        emitter = CordovaEventEmitter(commandDelegate: commandDelegate)
        labelModule = LabelCaptureModule(emitter: emitter)
        labelModule.didStart()
    }

    public override func dispose() {
        labelModule.didStop()
        emitter.removeCallbacks()
        super.dispose()
    }

    @objc(executeLabel:)
    func executeLabel(_ command: CDVInvokedUrlCommand) {
        guard let argsJson = command.defaultArgumentAsDictionary else {
            commandDelegate.send(
                .failure(with: "Invalid argument received in executeLabel"),
                callbackId: command.callbackId
            )
            return
        }

        let coreModuleName = String(describing: CoreModule.self)
        guard let coreModule = DefaultServiceLocator.shared.resolve(clazzName: coreModuleName) as? CoreModule else {
            commandDelegate.send(
                .failure(with: "Unable to retrieve the CoreModule from the locator."),
                callbackId: command.callbackId
            )
            return
        }

        let result = CordovaResult(commandDelegate, emitter: emitter, command: command)
        let handled = coreModule.execute(
            CordovaMethodCall(command: command),
            result: result,
            module: self.labelModule
        )

        if !handled {
            let methodName = argsJson["methodName"] as? String ?? "unknown"
            commandDelegate.send(.failure(with: "Unknown Core method: \(methodName)"), callbackId: command.callbackId)
        }
    }

    @objc(getDefaults:)
    func getDefaults(command: CDVInvokedUrlCommand) {
        let defaults = labelModule.getDefaults()
        let defaultsResult = ["LabelCapture": defaults]
        commandDelegate.send(
            .success(message: defaultsResult as CDVPluginResult.JSONMessage),
            callbackId: command.callbackId
        )
    }
}
